effekt/shared/src/main/scala/effekt/RecursiveDescent.scala

```diff
@@ -842,6 +842,24 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
       }
       left
 
+  inline def infixConfused(nonTerminal: () => Term, op: TokenKind, confused: TokenKind): Term =
+    nonterminal:
+      var left = nonTerminal()
+      while (peek.kind == op || peek.kind == confused) {
+        val opToken = {
+          val before = position
+          val tok = next()
+          val after = position
+
+          if (tok.kind == confused) softFail(s"`$confused` is not a logical operator, did you mean to use `$op`?", before, after)
+
+          Token(tok.start, tok.end, op) // hack
+        }
+        val right = nonTerminal()
+        left = binaryOp(left, opToken, right).withRangeOf(left, right)
+      }
+      left
```

Comment:
I'm not very happy that we need to duplicate the code here :(

The parser (and other parts of the compiler) are increasingly filled with details that are "only" relevant for error reporting.

Reply:
I'm also unhappy about the code duplication, hence the idea for the ☹️ combinator.

Reply:
On the other hand, parsing is, by itself, somewhat trivial, the hard part _is, and always has been_ recovery and error reporting. The current parser design is pretty good for mere parsing, but, in a way, much worse at any attempts at recovery or error reporting (we're missing a lot of context, labels, etc., _and_ we fail on the first error!)

Reply:
It's not that important for this exact mistake (`and` instead of `&&`) as I doubt it would deter anybody who can already code, but error recovery / usable errors are very important in the Effekt-specific bits (effects, caps, regions, captures).